README.md
## Introduction
--- 
**Notice: This repo has been hard forked into https://github.com/netfoundry/ebpf-tproxy-splicer for inclusion in 
CloudZiti edge-router builds as an optional interception mode. This repo will continue to be maintained for R&D.**
This is a project to develop an eBPF program that utilizes tc-bpf to act as a statefull ingress FW and to redirect 
ingress ipv4 udp/tcp flows toward dynamically created sockets that correspond to zero trust based services on OpenZiti
edge-routers. Those interested on how to setup an openziti development environment should visit 
https://github.com/openziti/ziti. 
Also note this is eBPF tc based so interception only occurs for traffic ingressing on the interface that the eBPF program
is attached to. To intercept packets generated locally by the router itself the eBPF program would need to be attached to
the loopback interface. The eBPF program also provides stateful inbound firewalling and only allows ssh, dhcp and arp
bypass by default. Initially the program will allow ssh to any address inbound however after the first tproxy mapping is
inserted by the map_update tool it will only allow ssh addressed to the IP address of the interface that tc has loaded the
eBPF program.  All other traffic must be configured as a service in an OpenZiti Controller which then informs the edge-router
which traffic flows to accept. The open ziti edge-router then uses the map_update user space app to insert rules to
allow traffic in on the interface tc is running on. For those interested in additional background on the project please visit: 
https://openziti.io/using-ebpf-tc-to-securely-mangle-packets-in-the-kernel-and-pass-them-to-my-secure-networking-application.  

A new addtion is firewall support for subtending devices for two interface scenarios i.e.
external and trusted.

    external inet <----> (ens33)[ebpf-router](ens37) <----> trusted clients

    with tproxy-splicer.o applied ingress on ens33 and oubound_track.o applied egress on ens33 the router will
    statefully track outbound udp and tcp connections on ens33 and allow the associated inbound traffic.  While
    running in this mode it does not make sense to add ziti tproxy rules and is meant for running as a traditional fw.
    As be for you can also create passthrough FW rules (set -t --tproxy-port to 0) which would also make sense in the mode for
    specific internet initiated traffic you might want to allow in.

    TCP:
        If the tcp connections close gracefully then the entries will remove upon connection closure. 
        if not then there is a 60 minute timeout that will remove the in active state if no traffic seen
        in either direction.

    UDP:
        State will remain active as long as packets tuples matching SRCIP/SPORT/DSTIP/DPORT are seen in
        either direction within 30 seconds.  If no packets seen in either dorection the state will expire.
        If an external packet enters the interface after expire the entry will be deleted.  if an egress
        packet fined a matching expired state it will return the state to active.

    In order to support this per interface rule awareness was added which allows each port range within a prefix
    to match a list of connected interfaces.  On a per interface basis you can decide to honor that list or not via
    a per-prefix-rules setting in the following manner via the map_update utility
    
    singly:
    ```
    sudo map_update -P <ifname>
    ```
    or 

    all interfaces:
    ```
    sudo map_update -P all
    ```

    In order to assign 1 to 3 interfaces to a rule you would use the new -N option in combination with the -I i.e.
    to associate the rule to end37 and lo:

    ```
    sudo map_update -I -c 172.16.31.0 -m 24 -l 443 -h 443 -t 44000 -p tcp -N ens37 -N lo
    ```

## Build
---
[To build from source. Click here!](./BUILD.md)

## Management After Deployment
---

### Attaching to interface

```bash   
sudo tc qdisc add dev <interface name>  clsact
sudo tc filter add dev <interface name> ingress bpf da obj tproxy_splicer.o sec action
sudo tc filter add dev <interface name> egress bpf da obj outbound_track.o sec action  //optional only if firewalling subtending devices 
sudo ufw allow in on <interface name> to any
```

ebpf will now take over firewalling this interface and only allow ssh, dhcp and arp till ziti
services are provisioned as inbound intercepts via the map_udate app. Router will statefully allow responses to router
initiated sockets as well. tc commands above do not survive reboot so would need to be added to startup service / script.

### Openziti Ingress

Testing with ziti-router after attaching. Build a ziti network first and create services as explained at [Host It Anywhere](https://docs.openziti.io/docs/learn/quickstarts/network/hosted/)

Grab the edge router binary `(>= v0.27.3)` at [open ziti](https://github.com/openziti/ziti/releases).

```bash
# Copy the user space map program to folder in $PATH i.e
sudo cp map_update /usr/bin

# In the router config.yml set tunnel mode to ebpf i.e.   
- binding: tunnel
  options:
    mode: tproxy:/path/to/map_update

# Run edge router command 
# Note: assumption - ziti-router and config.yml in PATH
sudo ziti-router run config.yml
```

### Detaching from interface:

```bash
sudo tc qdisc del dev <interface name>  clsact
```

## Ebpf Map User Space Management
---
Example: Insert map entry to direct SIP traffic destined for 172.16.240.0/24

```bash
Usage: ./map_update -I <ip dest address or prefix> -m <prefix length> -l <low_port> -h <high_port> -t <tproxy_port> -p <protocol>

sudo ./map_update -I -c 172.16.240.0 -m 24 -l 5060 -h 5060 -t 58997 -p udp
```

As mentioned earlier if you add -r, --route as argument the program will add 172.16.240.0/24 to the "lo" interface if it
does not overlap with an external LAN interface subnet.

Example: Disable ssh from interface.

This will disable default ssh action to pass to ip of local interface and then fall through to rule check instead where a more specific rule could
be applied.  This is a per interface setting and can be set for all interfaces except loopback.

```bash
Usage: ./map_update -x <interface-name> | all

sudo sudo ./map_update -x ens33
```


Example: Insert map entry to with source filteing to only allow rule for ip source 10.1.1.1/32.

```bash
Usage: ./map_update -I -c <ip dest address or prefix> -m <dest prefix len> -o <origin address or prefix> -n <origin prefix len> -l <low_port> -h <high_port> -t <tproxy_port> -p <protocol>

sudo sudo ./map_update -I -c 172.16.240.0 -m 24 -o 10.1.1.1 -n 32  -p tcp -l 22 -h 22 -t 0
```

Example: Insert FW rule for local router tcp listen port 443 where local router's tc interface ip address is 10.1.1.1
with tproxy_port set to 0 signifying local connect rule

```bash
sudo ./map_update -I -c 10.1.1.1 -m 32 -l 443 -h 443 -t 0 -p tcp  
```

Example: Monitor ebpf trace messages

```
sudo map_update -v all
sudo cat /sys/kernel/debug/tracing/trace_pipe
  
<idle>-0       [007] dNs.. 167940.070727: bpf_trace_printk: ens33
<idle>-0       [007] dNs.. 167940.070728: bpf_trace_printk: source_ip = 0xA010101
<idle>-0       [007] dNs.. 167940.070728: bpf_trace_printk: dest_ip = 0xAC10F001
<idle>-0       [007] dNs.. 167940.070729: bpf_trace_printk: protocol_id = 17
<idle>-0       [007] dNs.. 167940.070729: bpf_trace_printk: tproxy_mapping->5060 to 59423

<idle>-0       [007] dNs.. 167954.255414: bpf_trace_printk: ens33
<idle>-0       [007] dNs.. 167954.255414: bpf_trace_printk: source_ip = 0xA010101
<idle>-0       [007] dNs.. 167954.255415: bpf_trace_printk: dest_ip = 0xAC10F001
<idle>-0       [007] dNs.. 167954.255415: bpf_trace_printk: protocol_id = 6
<idle>-0       [007] dNs.. 167954.255416: bpf_trace_printk: tproxy_mapping->22 to 39839

```
Example: Remove previous entry from map
```bash
Usage: ./map_update -D -c <ip dest address or prefix> -m <prefix len> -l <low_port> -p <protocol>

sudo ./map_update -D -c 172.16.240.0 -m 24 -l 5060 -p udp
```

Example: Remove all entries from map
```
Usage: ./map_update -F

sudo ./map_update -F
```

Example: List all rules in map
```
Usage: ./map_update -L

sudo ./map_update -L

target     proto    origin              destination               mapping:                                                   interface list
------     -----    ---------------     ------------------        --------------------------------------------------------- ----------------
TPROXY     tcp      0.0.0.0/0           10.0.0.16/28              dpts=22:22                TPROXY redirect 127.0.0.1:33381  [ens33,lo]
TPROXY     tcp      0.0.0.0/0           10.0.0.16/28              dpts=30000:40000          TPROXY redirect 127.0.0.1:33381  []
TPROXY     udp      0.0.0.0/0           172.20.1.0/24             dpts=5000:10000           TPROXY redirect 127.0.0.1:59394  []
TPROXY     tcp      0.0.0.0/0           172.16.1.0/24             dpts=22:22                TPROXY redirect 127.0.0.1:33381  []
TPROXY     tcp      0.0.0.0/0           172.16.1.0/24             dpts=30000:40000          TPROXY redirect 127.0.0.1:33381  []
PASSTHRU   udp      0.0.0.0/0           192.168.3.0/24            dpts=5:7                  PASSTHRU to 192.168.3.0/24       []
PASSTHRU   udp      10.1.1.1/32         192.168.100.100/32        dpts=50000:60000          PASSTHRU to 192.168.100.100/32   []
PASSTHRU   tcp      10.230.40.1/32      192.168.100.100/32        dpts=60000:65535          PASSTHRU to 192.168.100.100/32   []
TPROXY     udp      0.0.0.0/0           192.168.0.3/32            dpts=5000:10000           TPROXY redirect 127.0.0.1:59394  []
PASSTHRU   tcp      0.0.0.0/0           192.168.100.100/32        dpts=60000:65535          PASSTHRU to 192.168.100.100/32   []
```
Example: List rules in map for a given prefix and protocol
```bash
# Usage: ./map_update -L -c <ip dest address or prefix> -m <prefix len> -p <protocol>
  
sudo map_update -L -c 192.168.100.100 -m 32 -p udp
  
target     proto    origin           destination              mapping:                                                  interface list
------     -----    --------         ------------------       --------------------------------------------------------- ------------------    
PASSTHRU   udp      0.0.0.0/0        192.168.100.100/32       dpts=50000:60000 	      PASSTHRU to 192.168.100.100/32     []
``` 

Example: List rules in map for a given prefix
```bash
# Usage: ./map_update -L -c <ip dest address or prefix> -m <prefix len> -p <protocol>

sudo map_update -L -c 192.168.100.100 -m 32

target     proto    origin           destination              mapping:                                                  interface list
------     -----    --------         ------------------       --------------------------------------------------------- -------------------
PASSTHRU   udp      0.0.0.0/0        192.168.100.100/32       dpts=50000:60000 	      PASSTHRU to 192.168.100.100/32     []
PASSTHRU   tcp      0.0.0.0/0        192.168.100.100/32       dpts=60000:65535	      PASSTHRU to 192.168.100.100/32     []
```

## Additional Distro testing
---

Fedora 36 kernel 6.0.5-200

On fedora I found that NetworkManager interferes with eBPF socket redirection and can
be unpredictable so belowis what I changed to get it working consistently. Other less
intrusive methods not requiring removal of NM might also be possible.

```bash
sudo yum install network-scripts
sudo systemctl enable network
sudo yum remove NetworkManager

sudo vi /etc/sysconfig/network-scripts/ifcfg-eth0

# if eth0 will be dhcp then something like:
BOOTPROTO=dhcp
DEVICE=eth0
ONBOOT=yes

# or if static
BOOTPROTO=static
IPADDR=192.168.61.70
NETMASK=255.255.255.0
DEVICE=eth1
ONBOOT=yes
```

The following grub change is only necessary on systems that do not use ethX naming by
default like vmware. this changes fedora back to using ethX for interface naming network-scripts looks for this nomenclature and will fail DHCP otherwise

```bash 
sudo vi /etc/default/grub

# change:
GRUB_CMDLINE_LINUX="rd.lvm.lv=fedora_fedora/root rhgb quiet"

# to:
GRUB_CMDLINE_LINUX="rd.lvm.lv=fedora_fedora/root rhgb quiet net.ifnames=0 biosdevname=0"

# then:
sudo grub2-mkconfig -o /boot/grub2/grub.cfg
```
This updates dhcp script to dynamically update systemd-resolved on per interface resolver

```bash
sudo vi /usr/sbin/dhclient-script

# change:
if [ -n "${new_domain_name_servers}" ]; then
    for nameserver in ${new_domain_name_servers} ; do
        echo "nameserver ${nameserver}" >> "${rscf}"
    done

# to:   
if [ -n "${new_domain_name_servers}" ]; then
    for nameserver in ${new_domain_name_servers} ; do
        echo "nameserver ${nameserver}" >> "${rscf}"
        systemd-resolve --interface "${interface}" --set-dns "${nameserver}"
    done
````

## Analisys
---

                                                        DIAGRAMS

![Diagram](packet-flow.drawio.png) 


## Deployment on Ubuntu
---
Since ebpf programs are not persistent over reboots, one needs a way to re-attach them to interfaces. We created a bash script to be run at boot time, and it is also triggered by ziti-router systemd service when it is restarted. Both are located under the files directory.
1. [tproxy_splicer_startup.sh](./files/scripts/tproxy_splicer_startup.sh)
1. [ziti-router.service](./files/services/ziti-router.service)

The script is created with comments, so hopefully the comments will help readers understand what is happening. One thing to pay attention to is `lanIf`. The `lanIf` option is configred in the ziti-router configuration file evey time either tproxy with iptables or ebpf is configured. 
```bash
# Here is the code snippet from the script:
MYIF=$(cat /opt/netfoundry/ziti/ziti-router/config.yml |yq '.listeners[] | select(.binding == "tunnel").options.lanIf')
```
As one can see we lookup lanIf (i.e. `.options.lanIf`) in the config.yml file located in the ziti-router directory. Based on that port everyting else is configured for the ebpf program to work. We also look up the fabric, edge, and health-check ports to update the tproxy map to allow these ports in along with `DNS UDP Port 53`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

---
## [0.2.10] - 2023-03-20

### Added

- Added option to disable ssh support on a per interface basis

### Fixed

- added additional check in tproxy_splicer_startup.sh to verify if icmp-echo is already enabled in UFW. To avoid having UFW 
  continually reload when ziti-router is restarting due to errors.

## [0.2.9] - 2023-03-20

### Fixed

- fixed etables insert performance issue when inserting large numbers of services. 

## [0.2.8] - 2023-03-20

### Added

- added -v, --verbose <ifname> option to enable and disable trace output.  -d --disable when added disables verbose mode.  default is disabled.

### Changed

- Updated the logic for the local fw rules in the lifecycle script; included source address in all rules as configured by the ufw firewall.

## [0.2.7] - 2023-03-15

### Fixed

- Updated the router service file with the single binary router run command
- Updated the ufw rule deletion logic, when diverter-disable is invoked

## [0.2.6] - 2023-03-13

### Added

- Added support for source address filtering
- Added rule to permananetly allow icmp echo replies from any source
- Added command line switch to etables to allow icmp echos into router
- Added check for maximum prefix tuples reached and current usage in etables -L

## [0.2.5] - 2023-02-21

### Bug Fix

- Added check for maximum port ranges reached for a particular prefix destination

## [0.2.4] - 2023-02-16

### Bug Fix

- Fixed typo in output when invalid IP address entered

## [0.2.3] - 2023-02-13

### Bug fix

- Fixed issue where if an interface exists that has an UNSPEC address family caused etables binary to segment fault.


## [0.2.2] - 2023-02-10

### Added

- 2 options to tproxy startup script
    - `--add-user-ingress-rules`        true if adding user ingress rules to ebpf map (based on yaml formatted content of ebpf_user_file=$EBPF_HOME/user_ingress_rules.yml)
    - `--delete-user-ingress-rules`     true if deleting user ingress rules from ebpf map (based on yaml formatted content of ebpf_user_file=$EBPF_HOME/user_ingress_rules.yml)

### Changed

- `--check-ebpf-status`  - now it indicates if the ebpf program is attached and shows the interface name in addtion to the trace raw command details.

---

## [0.2.1] - 2023-02-03

### Changed

- change map_update to etables

### Added 

- 2 options to etables program to further filter option -L --list
    - `-f, --passthrough`    list passthrough rules <optional list>
    - `-i, --intercepts`     list intercept rules <optional for list>

---

## [0.2.0] - 2023-01-31

### Changed

- diverter object file is now compiled with 3 different map size options
    - small  = 1000  entries
    - medium = 5000  entries
    - large  = 10000 entries

---

## [0.1.0] - 2023-01-30

### Added 

- Initial Version



files/scripts/tproxy_splicer_startup.sh
#!/bin/bash

# Help menu
function usage() 
{
    cat <<USAGE

    Usage: tproxy_splicer_startup.sh [--initial-setup] [--revert-tproxy] [--check-ebpf-status] [--add-user-ingress-rules] [--delete-user-ingress-rules]

    Options:
        --initial-setup                 true if setting up ebpf for first time
        --revert-tproxy                 true if reverting back to tproxy with iptables
        --check-ebpf-status             true if just checking the status of ebpf program
        --add-user-ingress-rules        true if adding user ingress rules to ebpf map
        --delete-user-ingress-rules     true if deleting user ingress rules from ebpf map
        -V, --version    current version
        -h, --help       help menu
USAGE
    exit 1
}

# Shows current version
function version()
{
    echo "0.1.0"
    exit 1
}

# Fucntion to check if ebpf program is attached if not re-attach it
attach_ebpf_program()
{
    # Look up the value of lanIf
    LANIF=$(cat $router_config_file |yq '.listeners[] | select(.binding == "tunnel").options.lanIf')
    # Check if lanIf is not empty
    if [ "$LANIF" ]; then
        # Then get IP
        LANIP=$(/sbin/ip add show "${LANIF}"|awk '$1=="inet" {print $2;}'|awk -v FS='/' '{print $1}')
        # Attach ebpf program to lanif interface
        ATTACHED=`eval /usr/sbin/tc filter show dev $LANIF ingress`
        # Check if tproxy splicer ebpf is attached 
        if [ $(echo  $ATTACHED | grep -w -c tproxy_splicer.o) == 1 ]; then  
            if [ $CHECK_EBPF_STATUS == false ] && [ $ADD_USER_INGRESS_RULES == false ] && [ $DELETE_USER_INGRESS_RULES == false ]; then
                echo "INFO: Found ebpf program tproxy_splicer and will re-add it to $LANIF now"
                /usr/sbin/tc qdisc del dev $LANIF clsact
                /usr/sbin/tc qdisc add dev $LANIF clsact
                /usr/sbin/tc filter add dev $LANIF ingress bpf da obj $tproxy_splicer_path sec action || { /usr/sbin/tc qdisc del dev $LANIF clsact; exit 1; }
                ufw allow in on $LANIF to any 
            else
                # Check if user ingress urle update was initiated
                if [ $ADD_USER_INGRESS_RULES == true ] || [ $DELETE_USER_INGRESS_RULES == true ]; then
                    update_map_user
                # Then it is status check
                else
                    echo "INFO: Ebpf is enabled; tproxy splicer is attached to $LANIF"
                    echo "TRACE: $ATTACHED"
                fi
            fi           
        else
            if [ $CHECK_EBPF_STATUS == false ] && [ $ADD_USER_INGRESS_RULES == false ] && [ $DELETE_USER_INGRESS_RULES == false ]; then
                echo "INFO: Not Found ebpf program tproxy_splicer and will add it to $LANIF now"
                /usr/sbin/tc qdisc add dev $LANIF clsact
                /usr/sbin/tc filter add dev $LANIF ingress bpf da obj $tproxy_splicer_path sec action || { /usr/sbin/tc qdisc del dev $LANIF clsact; exit 1; }
                ufw allow in on $LANIF to any
            else
                # Check if user ingress rule update was initiated
                if [ $ADD_USER_INGRESS_RULES == true ] || [ $DELETE_USER_INGRESS_RULES == true ]; then
                    update_map_user
                # Then it is status check
                else
                    echo "INFO: Ebpf is not enabled; tproxy splicer is not attached to $LANIF"
                fi
            fi
        fi
        if [ $CHECK_EBPF_STATUS == false ] && [ $ADD_USER_INGRESS_RULES == false ] && [ $DELETE_USER_INGRESS_RULES == false ]; then
            update_map_local
        fi
    else
        echo "ERROR: lanIf not found in $router_config_file, check the config file"
    fi
}

# Update tproxy map with access to local ports
function update_map_local() 
{
    # Save full path of map update into a variable
    if [ -f "$etables" ]; then
        # FW Ziti Fabric Rules based on router config
        ZT_FPORT=$(cat $router_config_file |yq '.link.listeners[] | select(.binding == "transport").advertise' |awk -v FS=':' '{print $3}')
        ZT_FTRANSPORT=$(cat $router_config_file |yq '.link.listeners[] | select(.binding == "transport").advertise' |awk -v FS=':' '{print $1}')
        if [[ -n $ZT_FPORT ]]; then
            ZT_F_RULES=$(sudo ufw status numbered | jc --ufw -p | jq -r --argjson PORT "$ZT_FPORT" '.rules | map(select(.to_ports == [$PORT]))')
            if [ "$ZT_FTRANSPORT" == "tls" ] && [[ -n $ZT_F_RULES ]]; then
                for rule in $(echo "${ZT_F_RULES}" | jq -r '.[] | @base64'); do
                    _jq() {
                        echo ${rule} | base64 --decode | jq -r ${1}
                    }
                    $etables -I -c $LANIP -m 32 -o $(_jq '.from_ip') -n $(_jq '.from_ip_prefix') -l $ZT_FPORT -h $ZT_FPORT -t 0 -p tcp
                done
            fi
        fi
        # FW Ziti Client Rules based on router config
        ZT_CPORT=$(cat $router_config_file |yq '.listeners[] | select(.binding == "edge").address' |awk -v FS=':' '{print $3}')
        ZT_CTRANSPORT=$(cat $router_config_file |yq '.listeners[] | select(.binding == "edge").address' |awk -v FS=':' '{print $1}')
        if [[ -n $ZT_CPORT ]]; then
            ZT_C_RULES=$(sudo ufw status numbered | jc --ufw -p | jq -r --argjson PORT "$ZT_CPORT" '.rules | map(select(.to_ports == [$PORT]))')
            if [ "$ZT_CTRANSPORT" == "tls" ] && [[ -n $ZT_C_RULES ]]; then
                for rule in $(echo "${ZT_C_RULES}" | jq -r '.[] | @base64'); do
                    _jq() {
                        echo ${rule} | base64 --decode | jq -r ${1}
                    }
                    $etables -I -c $LANIP -m 32 -o $(_jq '.from_ip') -n $(_jq '.from_ip_prefix') -l $ZT_CPORT -h $ZT_CPORT -t 0 -p tcp
                done
            fi
        fi
        # FW DNS Rules based on UFW Configuration
        DNS_RULES=$(sudo ufw status numbered | jc --ufw -p | jq -r '.rules | map(select(.to_ports == [53]))')
        if [[ -n $DNS_RULES ]]; then
            for rule in $(echo "${DNS_RULES}" | jq -r '.[] | @base64'); do
                _jq() {
                    echo ${rule} | base64 --decode | jq -r ${1}
                }
                $etables -I -c $LANIP -m 32 -o $(_jq '.from_ip') -n $(_jq '.from_ip_prefix') -l 53 -h 53 -t 0 -p $(_jq '.to_transport')
            done
        fi
        # FW Health-Check Rules for HC Port read from router config
        ZT_HCPORT=$(cat $router_config_file |yq '.web[] | select(.name == "health-check").bindPoints[].address' | awk -v FS=':' '{print $2}')
        if [[ -n $ZT_HCPORT ]]; then
            ZT_HC_RULES=$(sudo ufw status numbered | jc --ufw -p | jq -r --argjson PORT "$ZT_HCPORT" '.rules | map(select(.to_ports == [$PORT]))')
            if [[ -n $ZT_HC_RULES ]]; then
                for rule in $(echo "${ZT_HC_RULES}" | jq -r '.[] | @base64'); do
                    _jq() {
                        echo ${rule} | base64 --decode | jq -r ${1}
                    }
                    $etables -I -c $LANIP -m 32 -o $(_jq '.from_ip') -n $(_jq '.from_ip_prefix') -l $ZT_HCPORT -h $ZT_HCPORT -t 0 -p tcp
                done
            fi
        fi
        # invoke function to update ingress fw rules parsed from user file
        update_map_user
    fi
}

function update_map_user()
{
    # if user file exists
    if [ -f $ebpf_user_file ]; then
        if [[ -n  `yq '.rules' $ebpf_user_file` ]]; then
            RULES_NUM=`yq '.rules | keys | length' $ebpf_user_file`
            for ((i = 0 ; i < $RULES_NUM ; i++)); do 
                export ikey=$i
                PORTS=`yq '.rules[env(ikey)] | .ports' $ebpf_user_file`
                PROTOCOLS=`yq '.rules[env(ikey)] | .protocols' $ebpf_user_file`
                PROTO_LEN=`yq '.rules[env(ikey)] | .protocols | length' $ebpf_user_file`
                SOURCE_PREFIX=`yq '.rules[env(ikey)] | .source_prefix' $ebpf_user_file`
                for ((j = 0 ; j < $PROTO_LEN ; j++)); do 
                    export jkey=$j
                    if [[ -n $PORTS ]] && [[ -n $PROTOCOLS ]] && [[ -n $SOURCE_PREFIX ]]; then
                        LP=`yq '.rules[env(ikey)] | .ports[0]' $ebpf_user_file`
                        HP=`yq '.rules[env(ikey)] | .ports[1]' $ebpf_user_file`
                        P=`yq '.rules[env(ikey)] | .protocols[env(jkey)]' $ebpf_user_file`
                        SP=`yq '.rules[env(ikey)] | .source_prefix[0]' $ebpf_user_file`
                        SPL=`yq '.rules[env(ikey)] | .source_prefix[1]' $ebpf_user_file`
                        if [ $DELETE_USER_INGRESS_RULES == true ]; then
                            $etables -D -c $LANIP -m 32 -o $SP -n $SPL -l $LP -h $HP -t 0 -p $P
                        else 
                            $etables -I -c $LANIP -m 32 -o $SP -n $SPL -l $LP -h $HP -t 0 -p $P
                        fi
                    fi
                done
            done
            unset ikey
            unset jkey
        else
           echo "INFO: No rules found in $ebpf_user_file, nothing to do" 
        fi
        # Enable or disable icmp on $LANIF
        if [[ -n  `yq '.icmp' $ebpf_user_file` ]]; then
            ICMP=`yq '.icmp.enable' $ebpf_user_file`
            if [ $ICMP == true ]; then
                echo "INFO: ICMP is set to $ICMP"
                $etables -e $LANIF
		is_enabled=$(grep -r "\-A ufw-before\-input \-p icmp \-\-icmp\-type echo\-request \-j ACCEPT" /etc/ufw/before.rules)
		if [[ ${#is_enabled} -eq 0 ]]
		then
                    sed -i '/ufw-before-input.*icmp/s/DROP/ACCEPT/g' /etc/ufw/before.rules
                    ufw reload
		else
		    echo "ICMP:UFW already modified to accept icmp, nothing to do"
		fi
            else
                echo "INFO: ICMP is set to $ICMP"
                $etables -e $LANIF -d
                sed -i '/ufw-before-input.*icmp/s/ACCEPT/DROP/g' /etc/ufw/before.rules
                echo "WARNING: icmp disable will not take affect until after reboot"
            fi
	# Enable or disable ssh on $LANIF
        if [[ -n  `yq '.ssh' $ebpf_user_file` ]]; then
            SSH=`yq '.ssh.disable' $ebpf_user_file`
            if [ $SSH == true ]; then
                echo "INFO: ssh disabled is set to $SSH"
                $etables -x $LANIF
            else
                echo "INFO: ssh disabled is set to $SSH"
                $etables -x $LANIF -d
            fi
        else
            echo "INFO: No ssh found in $ebpf_user_file, nothing to do"
        fi
        else
            echo "INFO: No icmp found in $ebpf_user_file, nothing to do"
        fi
    else
        echo "INFO: File $ebpf_user_file not found, nothing to do"
    fi
}

################
##### Main #####
################

# Process the input options.
# Get Input Flags

INITIAL_SETUP=false
REVERT_TPROXY=false
CHECK_EBPF_STATUS=false
ADD_USER_INGRESS_RULES=false
DELETE_USER_INGRESS_RULES=false

while [ "$1" != "" ]; do
    case $1 in
    --initial-setup)
        INITIAL_SETUP=true;;
    --revert-tproxy)
        REVERT_TPROXY=true;;
    --check-ebpf-status)
        CHECK_EBPF_STATUS=true;;
    --add-user-ingress-rules)
        ADD_USER_INGRESS_RULES=true;;
    --delete-user-ingress-rules)
        DELETE_USER_INGRESS_RULES=true;;
    -h | --help)
        usage;;
    -V | --version)
        version;;
    *)
        usage;;
    esac
    shift
done

# get local ip for lanIf
# First check if the edge router config file exists
CLOUD_ZITI_HOME="/opt/netfoundry"
EBPF_HOME="$CLOUD_ZITI_HOME/ebpf"
router_config_file="$CLOUD_ZITI_HOME/ziti/ziti-router/config.yml"
etables="$EBPF_HOME/objects/etables"
tproxy_splicer_path="$EBPF_HOME/objects/tproxy_splicer.o"
zt_router_service_path="$EBPF_HOME/services/ziti-router.service"
ebpf_user_file="$EBPF_HOME/user_ingress_rules.yml" 
ebpf_map_home="/sys/fs/bpf/tc/globals"
# Is router config file present?
if [ -f "$router_config_file" ]; then
    # Look up tunnel edge binding
    TUNNEL=$(cat $CLOUD_ZITI_HOME/ziti/ziti-router/config.yml |yq '.listeners[] | select(.binding == "tunnel")')
    if [ "$TUNNEL" ]; then
        # Get Tunnel Mode
        TMODE=$(cat $CLOUD_ZITI_HOME/ziti/ziti-router/config.yml |yq '.listeners[] | select(.binding == "tunnel").options.mode')
        # If tproxy is set and initial set flag is set, both true?
        if [ "$TMODE" == null ] || [ "$TMODE" == "tproxy" ]; then
            if [ $INITIAL_SETUP == true ] ; then
                echo "INFO: Checking if diverter installed"
                if [ ! -f $tproxy_splicer_path ] || [ ! -f $zt_router_service_path ]; then
                    echo "ERROR: Diverter not installed, please run diverter-update command to install it"
                else
                    echo "INFO: Initial setup starting"
                    attach_ebpf_program
                    # set it to env var
                    export etables
                    yq -i '(.listeners[] | select(.binding == "tunnel").options | .mode) = "tproxy:"+strenv(etables)' $router_config_file 
                    # unset it to env var
                    unset etables
                    cp $EBPF_HOME/services/ziti-router.service /etc/systemd/system/
                    /usr/bin/systemctl daemon-reload 
                    /usr/bin/systemctl restart ziti-router.service
                fi
            else
                echo "INFO: MODE $TMODE is configured and initial setup flag is $INITIAL_SETUP. Nothing to do"
                if [ $CHECK_EBPF_STATUS == true ]; then
                    echo "INFO: Ebpf is not enabled; tproxy splicer is not attached to LANIF"
                fi
            fi 
        else 
            # Check if tproxy is already set for ebpf mode by evaluating value of mode furthur
            tproxy_ebpf_path=`echo $TMODE | awk -v FS=':' '{print $2}'`
            if [ "$tproxy_ebpf_path" == "$etables" ]; then
                if [ $REVERT_TPROXY == true ]; then
                    echo "INFO: Reverting back to tproxy iptables"
                    # Look up the value of lanIf
                    LANIF=$(cat $router_config_file |yq '.listeners[] | select(.binding == "tunnel").options.lanIf')
                    # Check if lanIf is not empty
                    if [ "$LANIF" ]; then
                        yq -i '(.listeners[] | select(.binding == "tunnel").options | .mode) = "tproxy"' $router_config_file 
                        /usr/sbin/tc qdisc del dev $LANIF clsact
                        /usr/bin/rm $ebpf_map_home/zt_tproxy_map
                        /usr/bin/rm $ebpf_map_home/ifindex_ip_map
                        /usr/bin/rm $ebpf_map_home/matched_map
                        /usr/bin/rm $ebpf_map_home/prog_map
                        /usr/bin/rm $ebpf_map_home/diag_map
                        /usr/bin/rm $ebpf_map_home/tuple_count_map
                        # delete ufw rule associated with ebpf
                        ufw_rule_num=$(sudo ufw status numbered | jc --ufw -p | jq -r --arg LANIF "$LANIF" '.rules[] | select(.to_interface == $LANIF).index')
                        while [[ $ufw_rule_num ]]; do
                            echo $ufw_rule_num
                            sudo ufw --force delete ${ufw_rule_num[0]}
                            ufw_rule_num=$(sudo ufw status numbered | jc --ufw -p | jq -r --arg LANIF "$LANIF" '.rules[] | select(.to_interface == $LANIF).index')
                        done
                        # Restart ziti router service after reverting all changes
                        /usr/bin/systemctl restart ziti-router.service
                    else
                        echo "ERROR: lanIf value is missing, could not revert config back to tproxy with iptables"
                    fi
                else
                    # Otherwise just run the main ebpf function                        
                    attach_ebpf_program
                fi
            else
                echo "INFO: TMODE $TMODE, nothing to do"
            fi
        fi
    else
        echo "INFO: tunnel mode not set, nothing to do"
    fi
else
    echo "INFO: router config file not found, perhaps it is expected!?"
fi

